README.md
diffiv
======

Fast implementation of Gandhi and Houde (2016) style instruments (A3 version). Available for: Linux, OSX [![Travis Build Status](https://api.travis-ci.org/mcaceresb/stata-diffiv.svg?branch=master)](https://travis-ci.org/mcaceresb/stata-diffiv).

`version 0.3.1 09Sep2018`

Installation
------------

I only have access to Stata 13.1, so I impose that to be the minimum.
```stata
local github "https://raw.githubusercontent.com"
net install diffiv, from(`github'/mcaceresb/stata-diffiv/master/build/)
* adoupdate, update
* ado uninstall diffiv
```

Usage
-----

```stata
sysuse auto
egen j = group(make)
expand 3, gen(t)
replace price = price + 10 * rnormal()
diffiv priceIV = price, market(t) good(j)
diffiv priceIV mpgIV = price mpg if foreign, market(t) good(j) replace
```

Compiling
---------

To compile, you will need

- The GNU compiler collection (gcc)
- git

Then you can run
```sh
cd ../stata-diffiv
make
```

License
-------

stata-diffiv is [MIT-licensed](https://github.com/mcaceresb/stata-diffiv/blob/master/LICENSE)


changelog.md
Change Log
==========

## diffiv-0.3.1 (2018-08-09)

### Bug fixes

- Corrected OS checking code.
- Corrected empty checking code.

## diffiv-0.3.0 (2018-08-09)

### Features

- `diffiv` for OSX now available.

### Bug fixes

- `diffiv` with multiple threshold variables should work correctly.

## diffiv-0.2.0 (2018-08-04)

### Features

- `diffiv` is a much faster version using C plugins.

## diffiv-0.1.0 (2018-07-31)

### Features

- `diffiv_mata` generates the instruments proposed by Gandhi and Houde (2016).
- Currently only the A3 version is available.


src/diffiv.c
/*********************************************************************
 * Program: diffiv.c
 * Blame:   Mauricio Caceres Bravo <mauricio.caceres.bravo@gmail.com>
 * Created: Wed Aug  1 09:12:03 EDT 2018
 * Updated: Tue Sep  4 17:20:51 EDT 2018
 * Purpose: Stata plugin for Gandhi and Houde (2016) style instruments.
 * Note:    See stata.com/plugins for more on Stata plugins
 * Version: 0.1.0
 *********************************************************************/

/**
 * @file diffiv.c
 * @author Mauricio Caceres Bravo
 * @date 1 Sep 2018
 * @brief Stata plugin
 *
 * This file should only ever be called from diffiv.ado
 *
 * @see help diffiv
 * @see http://www.stata.com/plugins for more on Stata plugins
 */

#include "diffiv.h"

int main()
{
    return(0);
}

int WinMain()
{
    return(0);
}

STDLL stata_call(int argc, char *argv[])
{
    ST_retcode rc = 0;
    setlocale(LC_ALL, "");
    struct StataInfo *st_info = malloc(sizeof(*st_info));
    st_info->free = 0;
    clock_t timer = clock();

    if ( (rc = ssf_parse_info (st_info)) ) {
        goto exit;
    }
    if ( st_info->benchmark )
        sf_running_timer (&timer, "\tPlugin step 1: Parse stata info");

    if ( (rc = ssf_read_varlist (st_info)) ) {
        goto exit;
    }
    if ( st_info->benchmark )
        sf_running_timer (&timer, "\tPlugin step 2: Read in variables");

    if ( (rc = ssf_diffiv (st_info)) ) {
        goto exit;
    }
    if ( st_info->benchmark )
        sf_running_timer (&timer, "\tPlugin step 3: Computed pairwise differences");

    if ( (rc = ssf_write_varlist (st_info)) ) {
        goto exit;
    }
    if ( st_info->benchmark )
        sf_running_timer (&timer, "\tPlugin step 4: Copied differences back to Stata");

exit:
    ssf_free (st_info);
    free (st_info);
    return (rc);
}

/**
 * @brief Parse variable info from Stata
 *
 * @param st_info Pointer to container structure for Stata info
 * @return Stores in @st_info various info from Stata for the pugin run
 */
ST_retcode ssf_parse_info (struct StataInfo *st_info)
{
    ST_retcode rc = 0;
    GT_size i, ix, in1, in2, N;
    GT_size k, benchmark;

    in1 = SF_in1();
    in2 = SF_in2();
    N   = in2 - in1 + 1;
    ix  = 0;

    if ( (rc = sf_scalar_size("diffiv_k", &k) )) goto exit;
    if ( (rc = sf_scalar_size("diffiv_benchmark", &benchmark) )) goto exit;

    st_info->index = calloc(N, sizeof st_info->index);
    st_info->empty = calloc(k, sizeof st_info->empty);

    if ( st_info->index == NULL ) return (sf_oom_error("ssf_parse_info", "index"));
    if ( st_info->empty == NULL ) return (sf_oom_error("ssf_parse_info", "empty"));

    st_info->free = 1;

    if ( (rc = sf_get_vector_bool ("diffiv_empty", st_info->empty) )) goto exit;

    for (i = 0; i < N; i++) {
        if ( SF_ifobs(i + in1) ) {
            st_info->index[ix] = i;
            ix++;
        }
    }

    st_info->in1 = in1;
    st_info->in2 = in2;
    st_info->N   = ix;

    st_info->k = k;
    st_info->benchmark = benchmark;

exit:
    return(rc);
}

/**
 * @brief Read varlist to hash from stata
 *
 * @param st_info Pointer to container structure for Stata info
 * @return Stores in @st_info->st_charx the input variables
 */
ST_retcode ssf_read_varlist (struct StataInfo *st_info)
{
    ST_retcode rc = 0;
    ST_double z;
    GT_size g, i, j, k, sel, ng, nempty, pos;

    st_info->grp = calloc(st_info->N, sizeof st_info->grp);
    st_info->K   = calloc(st_info->N * st_info->k, sizeof st_info->K);
    st_info->X   = calloc(st_info->N * st_info->k, sizeof st_info->X);
    st_info->Z   = calloc(st_info->N * st_info->k, sizeof st_info->Z);

    if ( st_info->grp == NULL ) return (sf_oom_error("ssf_read_varlist", "grp"));
    if ( st_info->K   == NULL ) return (sf_oom_error("ssf_read_varlist", "K"));
    if ( st_info->X   == NULL ) return (sf_oom_error("ssf_read_varlist", "X"));
    if ( st_info->Z   == NULL ) return (sf_oom_error("ssf_read_varlist", "Z"));

    st_info->free = 2;

    // Read in variables
    for (i = 0; i < st_info->N; i++) {
        sel = st_info->index[i] + st_info->in1;

        // First variable is grp
        if ( (rc = SF_vdata(1, sel, &z)) ) goto exit;
        st_info->grp[i] = (GT_bool) z;

        // Next k are thresholds
        for (k = 0; k < st_info->k; k++) {
            j = k + st_info->k * i;
            if ( (rc = SF_vdata(2 + k, sel, st_info->K + j)) ) goto exit;
        }

        // Next k are sources
        for (k = 0; k < st_info->k; k++) {
            j = k + st_info->k * i;
            if ( (rc = SF_vdata(2 + st_info->k + k, sel, st_info->X + j)) ) goto exit;
        }
    }

    // Initialize IV
    for (i = 0; i < st_info->N; i++) {
        for (k = 0; k < st_info->k; k++) {
            j = k + st_info->k * i;
            st_info->Z[j] = 0;
        }
    }

    // Empty if applicable
    nempty = 0;
    for (k = 0; k < st_info->k; k++) {
        nempty += st_info->empty[k];
    }

    if ( nempty ) {
        pos = 2 + st_info->k + st_info->k;
        for (i = 0; i < SF_nobs(); i++) {
            for (k = 0; k < st_info->k; k++) {
                if ( st_info->empty[k] ) {
                    if ( (rc = SF_vstore(pos + k, i + 1, SV_missval)) ) goto exit;
                }
            }
        }
    }

    // Allocate memory for groups
    st_info->ginfo = sf_panelsetup(st_info->grp, st_info->N, &(st_info->G));
    if ( st_info->ginfo == NULL ) return (sf_oom_error("ssf_read_varlist", "ginfo"));
    st_info->free = 3;

    // Largest group
    st_info->ng_max = 0;
    for (g = 0; g < st_info->G; g++) {
        ng = st_info->ginfo[g + 1] - st_info->ginfo[g];
        if ( ng > st_info->ng_max ) st_info->ng_max = ng;
    }

exit:
    return(rc);
}

/**
 * @brief Compute pairwise differences
 *
 * @param st_info Pointer to container structure for Stata info
 * @return Computes pairwise differences in Z
 */
ST_retcode ssf_diffiv (struct StataInfo *st_info)
{
    ST_double xi, xii, s;
    GT_size i, ii, j, k, seli, selii, starts, ends;

    for (j = 0; j < st_info->G; j++) {
        starts = st_info->ginfo[j];
        ends   = st_info->ginfo[j + 1];
        for (i = starts; i < ends; i++) {
            for (ii = starts; ii < ends; ii++) {
                for (k = 0; k < st_info->k; k++) {
                    seli  = k + st_info->k * i;
                    selii = k + st_info->k * ii;
                    s     = st_info->K[seli];
                    xi    = st_info->X[seli];
                    xii   = st_info->X[selii];
                    if ( fabs(xi - xii) <= s ) {
                        st_info->Z[seli] += xii;
                    }
                }
            }
            for (k = 0; k < st_info->k; k++) {
                seli = k + st_info->k * i;
                st_info->Z[seli] -= st_info->X[seli];
            }
        }
    }

    return(0);
}

/**
 * @brief Write instruments back to stata
 *
 * @param st_info Pointer to container structure for Stata info
 * @return Stores in stata the difference IV
 */
ST_retcode ssf_write_varlist (struct StataInfo *st_info)
{
    ST_retcode rc = 0;
    GT_size i, j, k, sel, pos;
    pos = 2 + st_info->k + st_info->k;
    for (i = 0; i < st_info->N; i++) {
        sel = st_info->index[i] + st_info->in1;
        for (k = 0; k < st_info->k; k++) {
            j = k + st_info->k * i;
            if ( (rc = SF_vstore(pos + k, sel, st_info->Z[j])) ) goto exit;
        }
    }

exit:
    return(rc);
}

/**
 * @brief Clean up st_info
 *
 * @param st_info Pointer to container structure for Stata info
 * @return Frees memory allocated to st_info objects
 */
void ssf_free (struct StataInfo *st_info)
{
    if ( st_info->free >= 1 ) {
        free (st_info->index);
    }
    if ( st_info->free >= 2 ) {
        free (st_info->grp);
        free (st_info->K);
        free (st_info->X);
        free (st_info->Z);
    }
    if ( st_info->free >= 3 ) {
        free (st_info->ginfo);
    }
}


src/diffiv.h
#ifndef DIFFIV_H
#define DIFFIV_H    1

#include <locale.h>
#include <math.h>
#include <string.h>
#include <unistd.h>

#include "difftools.c"

// Container structure for Stata-provided info
struct StataInfo {
    GT_size in1;
    GT_size in2;
    GT_size N;
    GT_size Nread;
    GT_size G;
    GT_size ng_max;
    GT_size free;
    //
    GT_size k;
    GT_bool benchmark;
    //
    GT_bool *empty;
    GT_size *index;
    GT_size *ginfo;
    GT_bool *grp;
    //
    ST_double *K;
    ST_double *X;
    ST_double *Z;
};

// Main functions
ST_retcode ssf_parse_info    (struct StataInfo *st_info);
ST_retcode ssf_read_varlist  (struct StataInfo *st_info);
ST_retcode ssf_diffiv        (struct StataInfo *st_info);
ST_retcode ssf_write_varlist (struct StataInfo *st_info);
void ssf_free (struct StataInfo *st_info);

#endif /* diffiv.h */


src/difftools.c
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include "gttypes.h"
#include "spi/stplugin.h"
#include "difftools.h"

/**
 * @brief Short wrapper to print to Stata
 *
 * Basic wrapper to print formatted strings to Stata
 *
 * @param *fmt a string to format
 * @param ... Arguments to pass to pritnf
 * @return Prints to Stata's console
 */
void sf_printf (const char *fmt, ...)
{
    va_list args;
    va_start (args, fmt);
    char buf[BUF_MAX];
    vsprintf (buf, fmt, args);
    SF_display (buf);
    // printf (buf);
    va_end (args);
}

/**
 * @brief Short wrapper to print error to Stata
 *
 * Basic wrapper to print formatted error strings to Stata
 *
 * @param *fmt a string to format
 * @param ... Arguments to pass to pritnf
 * @return Prints to Stata's console
 */
void sf_errprintf (const char *fmt, ...)
{
    va_list args;
    va_start (args, fmt);
    char buf[BUF_MAX];
    vsprintf (buf, fmt, args);
    SF_error (buf);
    va_end (args);
}

/**
 * @brief Update a running timer and print a message to satata console
 *
 * Prints a messasge to Stata that the running timer @timer was last set
 * @diff seconds ago. It then updates the timer to the current time.
 *
 * @param timer clock object containing time since last udpate
 * @param msg message to print before # of seconds
 * @return Print time since last update to Stata console
 */
void sf_running_timer (clock_t *timer, const char *msg)
{
    double diff  = (double) (clock() - *timer) / CLOCKS_PER_SEC;
    sf_printf (msg);
    sf_printf ("; %.3f seconds.\n", diff);
    *timer = clock();
}

/**
 * @brief Read scalar into unsigned integer
 *
 * @param st_scalar name of Stata scalar
 * @param sval Scalar value
 * @return Read scalar into GT_size variable
 */
ST_retcode sf_scalar_size (char *st_scalar, GT_size *sval)
{
    ST_double _double;
    ST_retcode rc = SF_scal_use(st_scalar, &_double);
    if ( rc == 0 ) {
        *sval = (GT_size) _double;
    }
    return (rc);
}

/**
 * @brief Panel setup based on vector
 */
ST_retcode sf_oom_error (char *step_desc, char *obj_desc)
{
    sf_errprintf ("%s: Unable to allocate memory for object '%s'.\n", step_desc, obj_desc);
    return (1702);
}

/**
 * @brief Set up variables for panel hashes
 *
 * Using an index array, generate info array with start and
 * ending positions of each group in the sorted vector.
 *
 * @param group 0/1 array with the start of each group
 * @J Number of groups
 *
 * @return info arary with start and end positions of each group
 */
GT_size * sf_panelsetup(GT_bool *group, GT_size N, GT_size *G)
{
    GT_size i = 0, j = 0;
    GT_size *ginfo;
    GT_bool *gptr;

    *G = 0;
    for (gptr = group; gptr < group + N; gptr++)
        *G += *gptr;

    ginfo = calloc(*G + 1, sizeof ginfo);
    if ( ginfo == NULL ) {
        return (NULL);
    }
    else {
        for (gptr = group; gptr < group + N; gptr++, i++) {
            if ( *gptr ) ginfo[j++] = i;
        }
        ginfo[j] = N;
        return (ginfo);
    }
}

/**
 * @brief Parse stata vector into C array
 *
 * @param st_matrix name of stata matrix to get
 * @param v array where to store the vector
 * @return Store min and max of @x
 */
ST_retcode sf_get_vector_bool (char *st_matrix, GT_bool *v)
{
    ST_double z;
    ST_retcode rc = 0;

    GT_size i;
    GT_size ncol = SF_col(st_matrix);
    GT_size nrow = SF_row(st_matrix);
    if ( (ncol > 1) & (nrow > 1) ) {
        sf_errprintf("tried to read a "GT_size_cfmt" by "GT_size_cfmt" matrix into an array\n", nrow, ncol);
        return (198);
    }
    else if ( (ncol == 0) & (nrow == 0) ) {
        sf_errprintf("tried to read a "GT_size_cfmt" by "GT_size_cfmt" matrix into an array\n", nrow, ncol);
        return (198);
    }

    if ( ncol > 1 ) {
        for (i = 0; i < ncol; i++) {
            if ( (rc = SF_mat_el(st_matrix, 1, i + 1, &z)) ) return (rc);
            v[i] = (GT_bool) z;
        }
    }
    else {
        for (i = 0; i < nrow; i++) {
            if ( (rc = SF_mat_el(st_matrix, i + 1, 1, &z)) ) return (rc);
            v[i] = (GT_bool) z;
        }
    }

    return (rc);
}


lib/update_version.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from datetime import datetime
from re import search, sub, findall
from os import path

import argparse
parser = argparse.ArgumentParser()
parser.add_argument('set',
                    nargs    = '*',
                    type     = str,
                    metavar  = 'SET',
                    default  = "main",
                    help     = "Sets of files to update")
parser.add_argument('--major',
                    dest     = 'major',
                    action   = 'store_true',
                    help     = "Major version",
                    required = False)
parser.add_argument('--minor',
                    dest     = 'minor',
                    action   = 'store_true',
                    help     = "Minor version",
                    required = False)
parser.add_argument('--patch',
                    dest     = 'patch',
                    action   = 'store_true',
                    help     = "Patch version",
                    required = False)
parser.add_argument('--dry-run',
                    dest     = 'dry_run',
                    action   = 'store_true',
                    help     = "Dry run",
                    required = False)
args = vars(parser.parse_args())

if not args['major'] and not args['minor'] and not args['patch']:
    print("Nothing to do.")
    exit(0)
else:
    major = int(args['major'])
    minor = int(args['minor'])
    patch = int(args['patch'])

main = [path.join("src",  "stata.toc"),
        path.join("src",  "diffiv.pkg"),
        path.join("src",  "diffiv_mata.ado"),
        path.join("src",  "diffiv.ado"),
        path.join("docs", "diffiv.sthlp"),
        path.join("README.md")]
test = [path.join("src", "tests.do")]
plug = [path.join("src", "diffiv.c")]

todo   = []
callok = False

if "main" in args['set']:
    todo  += main
    callok = True

if "test" in args['set']:
    todo += test
    callok = True

if "plug" in args['set']:
    todo += plug
    callok = True

if "all" in args['set']:
    todo = main + test + plug
    callok = True

if not callok:
    msg = "Don't know '{0}'".format(', '.join(args['set']))
    print(msg + "; specify any of 'main, test, plug, all'.")
    print("Will ignore; updating main files only.")
else:
    print("Will update version in files:")

months = ["Jan",
          "Feb",
          "Mar",
          "Apr",
          "May",
          "Jun",
          "Jul",
          "Aug",
          "Sep",
          "Oct",
          "Nov",
          "Dec"]
remonths = "(" + '|'.join(months) + ")"

for fname in todo:
    print("\t" + fname)
    with open(fname, 'r') as fhandle:
        flines = fhandle.readlines()

    with open(fname, 'w') as fhandle:
        for line in flines:
            if search('^d.+Distribution.+(\d{8,8})', line):
                today = datetime.strftime(datetime.now(), "%Y%m%d")
                oline = sub("\d{8,8}", today, line)
                print("\t\t" + line)
                print("\t\t" + oline)
                if args['dry_run']:
                    fhandle.write(line)
                else:
                    fhandle.write(oline)

                continue

            v = search('(^v|[Vv]ersion).+(\d+\.?){3,3}', line)
            s = search('Stata version', line)
            if v and not s:
                try:
                    res = findall('(\d+)(\.| |$)', line)
                    new_major = int(res[0][0]) + major
                    new_minor = 0 if major else int(res[1][0]) + minor
                    new_patch = 0 if major or minor else int(res[2][0]) + patch
                    new = "{0}.{1}.{2}".format(new_major, new_minor, new_patch)
                    oline = sub('(\d+\.?)+', new, line, 1)
                    if search("\d+" + remonths + "\d\d+", line):
                        today_day   = datetime.strftime(datetime.now(), "%d")
                        today_month = datetime.strftime(datetime.now(), "%B")
                        today_year  = datetime.strftime(datetime.now(), "%Y")
                        today = today_day + today_month[:3] + today_year
                        oline = sub("\d+" + remonths + "\d\d+", today, oline)

                    print("\t\t" + line)
                    print("\t\t" + oline)
                    if args['dry_run']:
                        fhandle.write(line)
                    else:
                        fhandle.write(oline)
                except:
                    fhandle.write(line)
            else:
                fhandle.write(line)


lib/travis_build_push.sh
#!/bin/bash

if [[ $TRAVIS_OS_NAME == 'osx' ]]; then
    REPO=`git config remote.origin.url`
    SSH_REPO=${REPO/https:\/\/github.com\//git@github.com:}

    git config user.name "Travis CI"
    git config user.email "$COMMIT_AUTHOR_EMAIL"

    echo "Adding OSX files."
    git checkout develop
    git branch -D osx
    git checkout -B osx
	cp build/*plugin lib/plugin/
    git add -f build/*osx*plugin
    git add -f lib/plugin/*osx*plugin

    echo "Committing OSX files."
    git commit -m "[Travis] Add plugin output for OSX build"

    ID_RSA=lib/id_rsa_travis_diffiv
    openssl aes-256-cbc -K $encrypted_22e162ff729d_key -iv $encrypted_22e162ff729d_iv -in ${ID_RSA}.enc -out ${ID_RSA} -d
    chmod 600 ${ID_RSA}
    eval `ssh-agent -s`
    ssh-add ${ID_RSA}

    echo "Pushing OSX files."
    ssh-keyscan github.com >> ~/.ssh/known_hosts
    git push -f ${SSH_REPO} osx

    rm -f ${ID_RSA}

    echo "Done"
fi


build.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# ---------------------------------------------------------------------
# Program: build.py
# Author:  Mauricio Caceres Bravo <mauricio.caceres.bravo@gmail.com>
# Created: Fri Aug 31 23:10:09 EDT 2018
# Updated: Fri Aug 31 23:10:09 EDT 2018
# Purpose: Main build file for diffiv (copies contents into ./build and
#          puts a .zip file in ./releases)

from os import makedirs, path, linesep, chdir, system, remove, rename
from shutil import copy2, rmtree
from sys import platform
from tempfile import gettempdir
from zipfile import ZipFile
from re import search
import argparse

# ---------------------------------------------------------------------
# Aux programs

try:
    from shutil import which
except:
    def which(program):
        import os

        def is_exe(fpath):
            return path.isfile(fpath) and os.access(fpath, os.X_OK)

        fpath, fname = path.split(program)
        if fpath:
            if is_exe(program):
                return program
        else:
            for epath in os.environ["PATH"].split(os.pathsep):
                epath = epath.strip('"')
                exe_file = path.join(epath, program)
                if is_exe(exe_file):
                    return exe_file

        return None


def makedirs_safe(directory):
    try:
        makedirs(directory)
        return directory
    except OSError:
        if not path.isdir(directory):
            raise


# ---------------------------------------------------------------------
# Command line parsing

parser = argparse.ArgumentParser()
parser.add_argument('--stata',
                    nargs    = 1,
                    type     = str,
                    metavar  = 'STATA',
                    default  = None,
                    required = False,
                    help     = "Path to stata executable")
parser.add_argument('--stata-args',
                    nargs    = 1,
                    type     = str,
                    metavar  = 'STATA_ARGS',
                    default  = None,
                    required = False,
                    help     = "Arguments to pass to Stata executable")
parser.add_argument('--make-flags',
                    nargs    = 1,
                    type     = str,
                    metavar  = 'MAKE_FLAGS',
                    default  = None,
                    required = False,
                    help     = "Arguments to pass to make")
parser.add_argument('--clean',
                    dest     = 'clean',
                    action   = 'store_true',
                    help     = "Clean build",
                    required = False)
parser.add_argument('--replace',
                    dest     = 'replace',
                    action   = 'store_true',
                    help     = "Replace build",
                    required = False)
parser.add_argument('--test',
                    dest     = 'test',
                    action   = 'store_true',
                    help     = "Run tests",
                    required = False)
parser.add_argument('--windows',
                    dest     = 'windows',
                    action   = 'store_true',
                    help     = "Compile for Windows from Unix environment.",
                    required = False)
args = vars(parser.parse_args())

# ---------------------------------------------------------------------
# Relevant files

diffiv_ssc = [
    "diffiv_mata.ado",
    "diffiv.ado",
    "diffiv.sthlp"
]

diffiv_zip = [
    "changelog.md",
    "diffiv.pkg",
    "stata.toc"
] + diffiv_ssc

diffiv_build = diffiv_zip + [
    "tests.do"
]

# ---------------------------------------------------------------------
# Run the script

# Remove buld
# -----------

if args['clean']:
    print("Removing build files")
    for bfile in diffiv_build:
        try:
            remove(path.join("build", bfile))
            print("\tdeleted " + bfile)
        except:
            try:
                remove(path.join("build", "diffiv", bfile))
                print("\tdeleted " + bfile)
            except:
                print("\t" + bfile + " not found")

    rc = system("make clean")
    exit(0)

makedirs_safe(path.join("build", "diffiv"))
makedirs_safe("releases")

# Stata executable
# ----------------

# I don't have stata on my global path, so to make the script portable
# I make it look for my local executable when Stata is not found.
if args['stata'] is not None:
    statadir = path.abspath(".")
    stataexe = args['stata'][0]
    statargs = "-b do" if args['stata_args'] is None else args['stata_args'][0]
    statado  = '"{0}" {1}'.format(stataexe, statargs)
elif which("stata") is None:
    statadir = path.expanduser("~/.local/stata13")
    stataexe = path.join(statadir, "stata")
    statargs = "-b do" if args['stata_args'] is None else args['stata_args']
    statado  = '"{0}" {1}'.format(stataexe, statargs)
else:
    statadir = path.abspath(".")
    stataexe = 'stata'
    statargs = "-b do" if args['stata_args'] is None else args['stata_args']
    statado  = '"{0}" {1}'.format(stataexe, statargs)

# Temporary files
# ---------------

maindir   = path.dirname(path.realpath(__file__))
tmpdir    = gettempdir()
tmpupdate = path.join(tmpdir, ".update_diffiv.do")

# Compile plugin files
# --------------------

if platform in ["linux", "linux2", "win32", "cygwin", "darwin"]:
    print("Trying to compile plugins for -diffiv-")
    print("(note: this assumes you copied libssl.a and libcrypto.a to lib/windows)")
    make_flags = args['make_flags'][0] if args['make_flags'] is not None else ""
    rc = system("make all {0}".format(make_flags))
    print("Success!" if rc == 0 else "Failed.")
else:
    print("Don't know platform '{0}'; compile manually.".format(platform))
    exit(198)

print("")

# Copy files to ./build
# ---------------------

gdir = path.join("build", "diffiv")
copy2("changelog.md", gdir)

copy2(path.join("src",  "diffiv.pkg"),      gdir)
copy2(path.join("src",  "stata.toc"),       gdir)
copy2(path.join("src",  "tests.do"),        gdir)
copy2(path.join("src",  "tests.do"),        "build")
copy2(path.join("src",  "diffiv_mata.ado"), gdir)
copy2(path.join("src",  "diffiv.ado"),      gdir)
copy2(path.join("docs", "diffiv.sthlp"),    gdir)

# Copy files to .zip folder in ./releases
# ---------------------------------------

# Get stata version
with open(path.join("src", "diffiv.ado"), 'r') as f:
    line    = f.readline()
    version = search('(\d+\.?)+', line).group(0)

plugins = ["diffiv_unix.plugin",
           "diffiv_windows.plugin",
           "diffiv_macosx.plugin"]
plugbak = plugins[:]
for plug in plugbak:
    if not path.isfile(path.join("build", plug)):
        alt = path.join("lib", "plugin", plug)
        if path.isfile(alt):
            copy2(alt, "build")
        else:
            print("Could not find '{0}'".format(plug))

chdir("build")
print("Compressing build files for diffiv-{0}".format(version))
if rc == 0:
    diffiv_anyplug = False
    for plug in plugbak:
        if path.isfile(plug):
            diffiv_anyplug = True
            rename(path.join(plug), path.join("diffiv", plug))
        else:
            plugins.remove(plug)
            print("\t'{0}' not found; skipping.".format(plug))

    if not diffiv_anyplug:
        print("WARNING: Could not find plugins despite build exit with 0 status.")
        exit(-1)

    diffiv_zip += plugins
else:
    print("WARNING: Failed to build plugins. Will exit.")
    exit(-1)

outzip = path.join(maindir, "releases", "diffiv-latest.zip".format(version))
with ZipFile(outzip, 'w') as zf:
    for zfile in diffiv_zip:
        zf.write(path.join("diffiv", zfile))
        print("\t" + path.join("diffiv", zfile))
        rename(path.join("diffiv", zfile), zfile)

chdir(maindir)
rmtree(path.join("build", "diffiv"))

# Copy files to send to SSC
# -------------------------

print("")
print("Compressing build files for diffiv-ssc.zip")
if rc == 0:
    diffiv_ssc += plugins
else:
    print("WARNING: Failed to build plugins. Will exit.")
    exit(-1)

chdir("build")
outzip = path.join(maindir, "releases", "diffiv-ssc.zip")
with ZipFile(outzip, 'w') as zf:
    for zfile in diffiv_ssc:
        zf.write(zfile)
        print("\t" + zfile)

# Replace package in ~/ado/plus
# -----------------------------

chdir(maindir)
if args["replace"]:
    if which(stataexe):
        with open(tmpupdate, 'w') as f:
            f.write("global builddir {0}".format(path.join(maindir, "build")))
            f.write(linesep)
            f.write("cap net uninstall diffiv")
            f.write(linesep)
            f.write("net install diffiv, from($builddir) replace")
            f.write(linesep)

        chdir(statadir)
        system(statado + " " + tmpupdate)
        remove(tmpupdate)
        # print(linesep + "Replaced diffiv in ~/ado/plus")
        chdir(maindir)
    else:
        print("Could not find Stata executable '{0}'.".format(stataexe))
        exit(-1)

# Run tests
# ---------

if args['test']:
    print("Running tests (see build/tests.log for output)")
    chdir("build")
    system(statado + " tests.do")
    chdir(maindir)
